File: App/Backend/src/services/quiz_service.py
from sqlalchemy.orm import Session

from Backend.src.models.quiz import (
    QuestionOption,
    Quiz,
    QuizAttempt,
    QuizQuestion,
)
from Backend.src.repositories.lesson_repository import LessonRepository
from Backend.src.repositories.quiz_repository import QuizRepository
from Backend.src.repositories.student_repository import StudentRepository
from Backend.src.utils.logger import logger

# =========================================================
# QUIZ CRUD
# =========================================================

def get_quizzes_by_lesson(
    db: Session,
    lesson_id: int,
    published_only: bool = False
) -> list[Quiz]:
    return QuizRepository.get_by_lesson(db, lesson_id, published_only)


def get_quiz(db: Session, quiz_id: int) -> Quiz | None:
    return QuizRepository.get_by_id(db, quiz_id)


def create_quiz(db: Session, quiz_data: dict) -> Quiz:
    lesson = LessonRepository.get_by_id(db, quiz_data["lesson_id"])
    if not lesson:
        raise ValueError("Lesson does not exist")

    if quiz_data["passing_marks"] > quiz_data["max_marks"]:
        raise ValueError("Passing marks cannot exceed maximum marks")

    quiz = QuizRepository.create_quiz(db, quiz_data)
    logger.info(f"Quiz created: {quiz.quiz_id} for lesson {quiz.lesson_id}")
    return quiz


def update_quiz(db: Session, quiz_id: int, updated_data: dict) -> Quiz | None:
    quiz = QuizRepository.get_by_id(db, quiz_id)
    if not quiz:
        return None

    new_passing = updated_data.get("passing_marks", quiz.passing_marks)
    new_max = updated_data.get("max_marks", quiz.max_marks)
    if new_passing > new_max:
        raise ValueError("Passing marks cannot exceed maximum marks")

    updated = QuizRepository.update_quiz(db, quiz, updated_data)
    logger.info(f"Quiz updated: {quiz_id}")
    return updated


def delete_quiz(db: Session, quiz_id: int) -> Quiz | None:
    quiz = QuizRepository.get_by_id(db, quiz_id)
    if not quiz:
        return None

    QuizRepository.delete_quiz(db, quiz)
    logger.info(f"Quiz deleted: {quiz_id}")
    return quiz


# =========================================================
# QUESTION & OPTIONS
# =========================================================

def add_question_to_quiz(db: Session, quiz_id: int, question_data: dict) -> QuizQuestion:
    quiz = QuizRepository.get_by_id(db, quiz_id)
    if not quiz:
        raise ValueError("Quiz does not exist")

    options_data = question_data.pop("options", [])
    if not options_data:
        raise ValueError("A question must have at least one option")

    return QuizRepository.create_question_with_options(db, quiz_id, question_data, options_data)


def get_quiz_questions(db: Session, quiz_id: int) -> list[QuizQuestion]:
    return QuizRepository.get_questions_by_quiz(db, quiz_id)


def get_question_options(db: Session, question_id: int) -> list[QuestionOption]:
    return QuizRepository.get_options_by_question(db, question_id)


# =========================================================
# QUIZ ATTEMPTS & AUTO-GRADING ENGINE
# =========================================================

def start_quiz_attempt(db: Session, quiz_id: int, student_id: int) -> QuizAttempt:
    quiz = QuizRepository.get_by_id(db, quiz_id)
    if not quiz:
        raise ValueError("Quiz does not exist")
    if not quiz.is_published:
        raise ValueError("Quiz is not published yet")

    student = StudentRepository.get_by_id(db, student_id)
    if not student:
        raise ValueError("Student does not exist")

    existing_attempts = QuizRepository.count_student_attempts(db, quiz_id, student_id)
    if existing_attempts >= quiz.max_attempts:
        raise ValueError(f"Maximum attempt limit ({quiz.max_attempts}) reached for this quiz")

    attempt_number = existing_attempts + 1
    attempt = QuizRepository.create_attempt(db, quiz_id, student_id, attempt_number)
    logger.info(f"Student {student_id} started Quiz {quiz_id} Attempt #{attempt_number}")
    return attempt


def submit_quiz_attempt(
    db: Session,
    attempt_id: int,
    student_id: int,
    answers_data: list[dict]
) -> QuizAttempt:
    attempt = QuizRepository.get_attempt_by_id(db, attempt_id)
    if not attempt:
        raise ValueError("Quiz attempt not found")
    if attempt.student_id != student_id:
        raise ValueError("You can only submit your own quiz attempt")
    if attempt.status != "in_progress":
        raise ValueError("This quiz attempt has already been submitted")

    quiz = QuizRepository.get_by_id(db, attempt.quiz_id)
    if not quiz:
        raise ValueError("Quiz not found")

    total_marks_awarded = 0.0
    evaluated_answers = []

    for ans_item in answers_data:
        question_id = ans_item["question_id"]
        selected_option_id = ans_item.get("selected_option_id")

        question = QuizRepository.get_question_by_id(db, question_id)
        if not question:
            continue

        marks_for_this_q = 0.0
        if selected_option_id:
            option = QuizRepository.get_option_by_id(db, selected_option_id, question_id)
            if option and option.is_correct:
                marks_for_this_q = float(question.marks)

        total_marks_awarded += marks_for_this_q
        evaluated_answers.append({
            "question_id": question_id,
            "selected_option_id": selected_option_id,
            "marks_awarded": marks_for_this_q
        })

    passed = total_marks_awarded >= float(quiz.passing_marks)
    total_score = round(total_marks_awarded, 2)

    completed_attempt = QuizRepository.save_attempt_results(
        db=db,
        attempt=attempt,
        answers=evaluated_answers,
        total_marks=total_score,
        passed=passed
    )
    logger.info(
        f"Quiz Attempt {attempt_id} evaluated: Score {total_score}/{quiz.max_marks} (Passed: {passed})"
    )
    return completed_attempt


def get_student_quiz_attempts(
    db: Session,
    quiz_id: int,
    student_id: int
) -> list[QuizAttempt]:
    return QuizRepository.get_student_attempts(db, quiz_id, student_id)

File: App(SpringBoot)/src/main/java/com/Project/LearningManagementSystem/service/QuizService.java
package com.Project.LearningManagementSystem.service;

import com.Project.LearningManagementSystem.dto.QuizDtos.OptionCreateRequest;
import com.Project.LearningManagementSystem.dto.QuizDtos.OptionResponse;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuestionCreateRequest;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuestionResponse;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuizAttemptResponse;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuizCreateRequest;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuizDetailResponse;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuizResponse;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuizSubmitRequest;
import com.Project.LearningManagementSystem.dto.QuizDtos.QuizUpdateRequest;
import com.Project.LearningManagementSystem.dto.QuizDtos.StudentAnswerResponse;
import com.Project.LearningManagementSystem.dto.QuizDtos.SubmitAnswerItem;
import com.Project.LearningManagementSystem.entity.QuestionOption;
import com.Project.LearningManagementSystem.entity.Quiz;
import com.Project.LearningManagementSystem.entity.QuizAttempt;
import com.Project.LearningManagementSystem.entity.QuizQuestion;
import com.Project.LearningManagementSystem.entity.StudentAnswer;
import com.Project.LearningManagementSystem.exception.BadRequestException;
import com.Project.LearningManagementSystem.exception.ForbiddenException;
import com.Project.LearningManagementSystem.exception.ResourceNotFoundException;
import com.Project.LearningManagementSystem.repository.LessonRepository;
import com.Project.LearningManagementSystem.repository.QuestionOptionRepository;
import com.Project.LearningManagementSystem.repository.QuizAttemptRepository;
import com.Project.LearningManagementSystem.repository.QuizQuestionRepository;
import com.Project.LearningManagementSystem.repository.QuizRepository;
import com.Project.LearningManagementSystem.repository.StudentAnswerRepository;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class QuizService {

    private final QuizRepository quizRepository;
    private final QuizQuestionRepository questionRepository;
    private final QuestionOptionRepository optionRepository;
    private final QuizAttemptRepository attemptRepository;
    private final StudentAnswerRepository answerRepository;
    private final LessonRepository lessonRepository;

    public QuizResponse toQuizResponse(Quiz q) {
        return new QuizResponse(
                q.getQuizId(),
                q.getLessonId(),
                q.getTitle(),
                q.getDescription(),
                q.getMaxMarks(),
                q.getPassingMarks(),
                q.getDurationMinutes(),
                q.getMaxAttempts(),
                q.isPublished());
    }

    @Cacheable(value = "quizzes", key = "#courseId")
    public List<QuizResponse> getQuizzesByCourse(Integer courseId) {
        return quizRepository.findByCourseId(courseId).stream().map(this::toQuizResponse).toList();
    }

    public QuizDetailResponse getQuizDetail(Integer quizId, boolean isStudent) {
        Quiz q = quizRepository.findById(quizId)
                .orElseThrow(() -> new ResourceNotFoundException("Quiz not found: " + quizId));

        List<QuizQuestion> questions = questionRepository.findByQuizIdOrderByQuestionIdAsc(quizId);
        List<QuestionResponse> qResponses = new ArrayList<>();

        for (QuizQuestion quest : questions) {
            List<QuestionOption> options = optionRepository.findByQuestionId(quest.getQuestionId());
            List<OptionResponse> optResponses = options.stream().map(opt -> new OptionResponse(
                    opt.getOptionId(),
                    opt.getQuestionId(),
                    opt.getOptionText(),
                    isStudent ? null : opt.isCorrect())).toList();

            qResponses.add(new QuestionResponse(
                    quest.getQuestionId(),
                    quest.getQuizId(),
                    quest.getQuestionText(),
                    quest.getQuestionType(),
                    quest.getMarks(),
                    optResponses));
        }

        return new QuizDetailResponse(
                q.getQuizId(),
                q.getLessonId(),
                q.getTitle(),
                q.getDescription(),
                q.getMaxMarks(),
                q.getPassingMarks(),
                q.getDurationMinutes(),
                q.getMaxAttempts(),
                q.isPublished(),
                qResponses);
    }

    @Transactional
    @CacheEvict(value = "quizzes", allEntries = true)
    public QuizResponse createQuiz(QuizCreateRequest request) {
        lessonRepository.findById(request.getLesson_id())
                .orElseThrow(() -> new ResourceNotFoundException(
                        "Lesson not found: " + request.getLesson_id()));

        if (request.getPassing_marks().compareTo(request.getMax_marks()) > 0) {
            throw new BadRequestException("Passing marks cannot exceed maximum marks");
        }

        Quiz quiz = new Quiz();
        quiz.setLessonId(request.getLesson_id());
        quiz.setTitle(request.getTitle().trim());
        quiz.setDescription(request.getDescription());
        quiz.setMaxMarks(request.getMax_marks());
        quiz.setPassingMarks(request.getPassing_marks());
        quiz.setDurationMinutes(request.getDuration_minutes());
        quiz.setMaxAttempts(request.getMax_attempts());
        quiz.setPublished(request.is_published());
        quiz.setCreatedAt(LocalDateTime.now());
        quizRepository.save(quiz);

        return toQuizResponse(quiz);
    }

    @Transactional
    @CacheEvict(value = "quizzes", allEntries = true)
    public QuizResponse updateQuiz(Integer quizId, QuizUpdateRequest request) {
        Quiz quiz = quizRepository.findById(quizId)
                .orElseThrow(() -> new ResourceNotFoundException("Quiz not found: " + quizId));

        if (request.getTitle() != null && !request.getTitle().isBlank()) {
            quiz.setTitle(request.getTitle().trim());
        }
        if (request.getDescription() != null) {
            quiz.setDescription(request.getDescription());
        }
        if (request.getMax_marks() != null) {
            quiz.setMaxMarks(request.getMax_marks());
        }
        if (request.getPassing_marks() != null) {
            quiz.setPassingMarks(request.getPassing_marks());
        }
        if (quiz.getPassingMarks().compareTo(quiz.getMaxMarks()) > 0) {
            throw new BadRequestException("Passing marks cannot exceed maximum marks");
        }
        if (request.getDuration_minutes() != null) {
            quiz.setDurationMinutes(request.getDuration_minutes());
        }
        if (request.getMax_attempts() != null) {
            quiz.setMaxAttempts(request.getMax_attempts());
        }
        if (request.getIs_published() != null) {
            quiz.setPublished(request.getIs_published());
        }
        quiz.setUpdatedAt(LocalDateTime.now());
        quizRepository.save(quiz);

        return toQuizResponse(quiz);
    }

    @Transactional
    @CacheEvict(value = "quizzes", allEntries = true)
    public void deleteQuiz(Integer quizId) {
        if (!quizRepository.existsById(quizId)) {
            throw new ResourceNotFoundException("Quiz not found: " + quizId);
        }
        quizRepository.deleteById(quizId);
    }

    @Transactional
    @CacheEvict(value = "quizzes", allEntries = true)
    public QuestionResponse addQuestion(Integer quizId, QuestionCreateRequest request) {
        if (!quizRepository.existsById(quizId)) {
            throw new ResourceNotFoundException("Quiz not found: " + quizId);
        }
        if (request.getOptions() == null || request.getOptions().isEmpty()) {
            throw new BadRequestException("A question must have at least one option");
        }

        QuizQuestion question = new QuizQuestion();
        question.setQuizId(quizId);
        question.setQuestionText(request.getQuestion_text().trim());
        question.setQuestionType(request.getQuestion_type() != null ? request.getQuestion_type() : "mcq");
        question.setMarks(request.getMarks() != null ? request.getMarks() : BigDecimal.ONE);
        questionRepository.save(question);

        List<OptionResponse> optResponses = new ArrayList<>();
        for (OptionCreateRequest optReq : request.getOptions()) {
            QuestionOption opt = new QuestionOption();
            opt.setQuestionId(question.getQuestionId());
            opt.setOptionText(optReq.getOption_text().trim());
            opt.setCorrect(optReq.is_correct());
            optionRepository.save(opt);
            optResponses.add(
                    new OptionResponse(opt.getOptionId(), opt.getQuestionId(), opt.getOptionText(), opt.isCorrect()));
        }

        return new QuestionResponse(
                question.getQuestionId(),
                question.getQuizId(),
                question.getQuestionText(),
                question.getQuestionType(),
                question.getMarks(),
                optResponses);
    }

    @Transactional
    public QuizAttemptResponse startQuizAttempt(Integer quizId, Integer studentId) {
        Quiz quiz = quizRepository.findById(quizId)
                .orElseThrow(() -> new ResourceNotFoundException("Quiz not found: " + quizId));

        if (!quiz.isPublished()) {
            throw new BadRequestException("Quiz is not published yet");
        }

        List<QuizAttempt> attempts = attemptRepository.findByQuizIdAndStudentId(quizId, studentId);
        if (attempts.size() >= quiz.getMaxAttempts()) {
            throw new BadRequestException(
                    "Maximum attempt limit (" + quiz.getMaxAttempts() + ") reached for this quiz");
        }

        int attemptNumber = attempts.size() + 1;
        QuizAttempt attempt = new QuizAttempt();
        attempt.setQuizId(quizId);
        attempt.setStudentId(studentId);
        attempt.setAttemptNumber(attemptNumber);
        attempt.setStartedAt(LocalDateTime.now());
        attempt.setStatus("in_progress");
        attemptRepository.save(attempt);

        return toAttemptResponse(attempt);
    }

    @Transactional
    public QuizAttemptResponse submitQuizAttempt(Integer attemptId, Integer studentId, QuizSubmitRequest request) {
        QuizAttempt attempt = attemptRepository.findById(attemptId)
                .orElseThrow(() -> new ResourceNotFoundException("Quiz attempt not found: " + attemptId));

        if (!attempt.getStudentId().equals(studentId)) {
            throw new ForbiddenException("You can only submit your own quiz attempt");
        }
        if (!"in_progress".equals(attempt.getStatus())) {
            throw new BadRequestException("This quiz attempt has already been submitted");
        }

        Quiz quiz = quizRepository.findById(attempt.getQuizId())
                .orElseThrow(() -> new ResourceNotFoundException("Quiz not found"));

        BigDecimal totalScore = BigDecimal.ZERO;
        List<StudentAnswerResponse> answerResponses = new ArrayList<>();

        for (SubmitAnswerItem item : request.getAnswers()) {
            QuizQuestion question = questionRepository.findById(item.getQuestion_id()).orElse(null);
            if (question == null)
                continue;

            BigDecimal marksAwarded = BigDecimal.ZERO;
            if (item.getSelected_option_id() != null) {
                QuestionOption opt = optionRepository.findById(item.getSelected_option_id()).orElse(null);
                if (opt != null && opt.isCorrect() && opt.getQuestionId().equals(question.getQuestionId())) {
                    marksAwarded = question.getMarks();
                }
            }

            totalScore = totalScore.add(marksAwarded);

            StudentAnswer answer = new StudentAnswer();
            answer.setAttemptId(attemptId);
            answer.setQuestionId(question.getQuestionId());
            answer.setSelectedOptionId(item.getSelected_option_id());
            answer.setMarksAwarded(marksAwarded);
            answerRepository.save(answer);

            answerResponses.add(new StudentAnswerResponse(
                    answer.getAnswerId(), answer.getAttemptId(), answer.getQuestionId(), answer.getSelectedOptionId(),
                    answer.getMarksAwarded()));
        }

        boolean passed = totalScore.compareTo(quiz.getPassingMarks()) >= 0;
        attempt.setSubmittedAt(LocalDateTime.now());
        attempt.setMarks(totalScore);
        attempt.setStatus("completed");
        attempt.setPassed(passed);
        attemptRepository.save(attempt);

        QuizAttemptResponse resp = toAttemptResponse(attempt);
        resp.setAnswers(answerResponses);
        return resp;
    }

    public List<QuizAttemptResponse> getStudentAttempts(Integer quizId, Integer studentId) {
        return attemptRepository.findByQuizIdAndStudentId(quizId, studentId).stream().map(this::toAttemptResponse)
                .toList();
    }

    public QuizAttemptResponse toAttemptResponse(QuizAttempt a) {
        List<StudentAnswerResponse> answers = answerRepository.findByAttemptId(a.getAttemptId()).stream()
                .map(ans -> new StudentAnswerResponse(
                        ans.getAnswerId(), ans.getAttemptId(), ans.getQuestionId(), ans.getSelectedOptionId(),
                        ans.getMarksAwarded()))
                .toList();

        return new QuizAttemptResponse(
                a.getAttemptId(),
                a.getQuizId(),
                a.getStudentId(),
                a.getAttemptNumber(),
                a.getStartedAt(),
                a.getSubmittedAt(),
                a.getMarks(),
                a.getStatus(),
                a.getPassed(),
                answers);
    }
}
